### ISSUE TITLE
Allow tasks to be variadic

### ISSUE THREAD

**Issue by qiemem** [NetLogo/NetLogo#771]
Currently, if a variadic procedure is wrapped in a task, the task inherits the default arity of the procedure, but not its variadic nature. Practical upshot of this:

```
observer> show (map list [1 2 3] [4 5 6] [7 8 9])
observer: [[1 4] [2 5] [3 6]]
```
(note that `map` implicitly wrap procedures in tasks)

This prevents NetLogo from having, for instance, a variadic `zip` function (since `map list` only works as a binary zip, as shown above).

It would be great if the resulting tasks were variadic as well.

**Comment by SethTisue** [comment]
Perhaps some syntax for defining arbitrary variadic tasks could be provided as well.

**Comment by qiemem** [comment]
I wasn't sure if I should clump new syntax in together with this simple enhancement. But now that you mentioned it, some like `?*` to get he list of all arguments passed to a task would be cool.

**Comment by frankduncan** [comment]
This is starting to feel too magical.  I think we should have an alternate task syntax that has explicit variable names rather than continue to figure out what type the task is from what occurs in and around it

**Comment by SethTisue** [comment]
example?

**Comment by frankduncan** [comment]
lambda [ a b &rest foo ] [ body ] ?

I don't know, I guess the task stuff always seemed a little magical to me, and this moves it further in that direction.  For some of the netlogo stuff (like self and myself), magic seems appropriate to make the language feel a certain way, but this doesn't mesh.

Though of course javascript has arguments, so it's not like this idea is without precendent.

Should we give ?* to procedures that don't define arguments to allow variadic user procedures?

**Comment by qiemem** [comment]
So as not to derail this issue, I've created an issue for discussing named arguments: #773 773

**Comment by SethTisue** [comment]
I agree it would be weird if there were user-defined variadic tasks but not user-defined variadic procedures.

**Comment by mrerrormessage** [comment]
I think the core of this issue doesn't necessarily require much of a change - I would almost say that it's a bug rather than an enhancement. What's really going on here is that `map` - a variadic reporter - accepts `list` - another variadic reporter - and doesn't behave as though `list` is variadic. I did a quick review of the code and `map` and `list` both *seem* to do the right thing. There is a relevant [comment in ExpressionParser](https://github.com/NetLogo/NetLogo/blob/5.x/src/main/org/nlogo/compiler/ExpressionParser.scala#L366-L371) that I think helps us understand why it isn't parsed this way (since this seems to be primarily a parsing problem in assigning the wrong number of formals to `_list`). There seems to be a bit of (well-deserved) ambiguity when a variadic HOF accepts a variadic reporter because we can't be sure whether we should (a) apply the reporter task to the arguments and pass the result of that as the function (and to how many arguments we should apply it) or (b) pass in the variadic reporter as variadic and pass in the remainder of the arguments as themselves. I think this could be resolved in (at least in the case of built-in primitives) by examining the return type. Since we know that the `list` primitive can never return a reporter task, we can conclude that `list` must be the procedure being passed into `map`. In general, however, the problem seems intractable within NetLogo without some considerable work.

**Comment by SethTisue** [comment]
I agree that the original issue Bryan raised might not be _that_ hard to fix. It's nontrivial though; `nvm/Task.scala` has multiple occurrences of `formals: Array[api.Let]` and `formals.size`.

**Comment by mrerrormessage** [comment]
I definitely don't think it's trivial, just that it can be done without major structural changes, and certainly without the addition of variadic tasks. I also wanted to point out that there was a reason that this wasn't done originally. This also makes the parser/compiler "smarter." This could be a good thing from a certain point of view, but I can also see arguments that it would be a bad thing. Variadic tasks would certainly need the compiler to be smarter, but the addition of variadic tasks would allow the implementation of this to be less hack-y, in the sense that it is generalizable to arbitrary tasks.

**Comment by SethTisue** [comment]
Oh, I was assuming we wanted this to work:

    let x task list
    show (map x [1 2 3] [4 5 6] [7 8 9])

So I was assuming variadic tasks were necessary. If you only care about making Bryan's original example work, then maybe it's possible to handle it in the parser. ExpressionParser is really hairy, though, so I'm not sure I like the idea of hacking it to support this. Allowing tasks to actually be variadic seems cleaner?

**Comment by SethTisue** [comment]
Note that Bryan's original example has an easy workaround:

    (map [(list ?1 ?2 ?3)] [1 2 3] [4 5 6] [7 8 9])

so I'm not sure it's really worth fixing only this if there is no other benefit.

**Comment by mrerrormessage** [comment]
This is interesting. Three are two different problems here with two different solutions
1) The problem I was solving for (Bryan's original one). The problem here is that ExpressionParser is actually discarding relevant information when coercing variadic procedures into tasks. This problem came into being because it is difficult to know how to parse a variadic procedure applied to variadic procedure. There may be some workaround that may involve modifying ExpressionParser. I see this as a bug because you can get inconsistent behavior from built-in primitives. Others may see it differently.
2) The problem you brought up (variadic tasks, including those built from primitives). This problem is more easily solved within the parser because you can't mean to apply the task within the procedure call. However, solving that problem would almost certainly require some sort of fix/change within nvm. I'm not familiar enough with nvm to say where that would be.

**Comment by SethTisue** [comment]
Yeah, clear now.

**Comment by qiemem** [comment]
Awesome, thanks @mrerrormessage. I was hoping this issue would remain focused (thus the creation of #773). This issue was meant to be focused on maintaining the variadic behavior of primitives when they are wrapped in tasks.

I'm not sure I fully understand where the ambiguity comes in. As far as I can tell, the next token after a primitive that takes in a task is *always* wrapped in a task (except for `task` itself). More specifically, a sequence of tokens following a task-taking primitive will be parsed as a single expression if and only if they are surrounded by parentheses (or the first token is `task` in which case, you can drop `task`, and recheck this biconditional). Thus, `(map list [1] [2] [3])` is not ambiguous. Even `(map my-random-procedure [1] [2] [3])` is not ambiguous. What am I missing here?

BTW, I just used `map` as it was the most practical example I could think of. I agree with @SethTisue in that hacking a fix to that specific issue would be undesirable. That said, I don't think adding something like `?*` is necessary for this issue. Perhaps the better test case would be:

```
let my-task task list
(runresult my-task 1 2 3)
```

I confess, I don't understand the comment that @mrerrormessage refers to. How could `map (f a) ...` be misparsed? Isn't that unambiguously "evaluate `(f a)` and pass the result to `map`"?

**Comment by mrerrormessage** [comment]
@qiemem , it seems hard to misparse from the perspective of a code author, but from the compiler's perspective, the type signatures (we don't call them that in NetLogo, but that's effectively what they are). 

Consider the types here (in pseudo-scala)
```scala
def map[A, B](f: A => B, xs: A*): List[B]
def list[A](xs: A*): List[A]
def variadicProcedure[A, B](xs: A*): B
```
so `map list x y z` is easy to determine from the type signatures.

Now consider what happens when variadicProcedure is of type `A* => (B => C)`. In that case it is not decidable without full parenthesization which arguments variadicProcedure is applied to, and which map is applied to. For instance, suppose I have a procedure `filter-lists` which takes any number of elements and returns a reporter that reports true if the element was one of the elements passed into `filter-lists` and false otherwise. It is of type `Elem* => (Elem => Boolean)`. Now `map filter-list 1 2 3 4 5 6` could be `(map (filter-list) 1 2 3 4 5 6)`, `(map (filter-list 1 2 3) 4 5 6)`, or `(map (filter-list 1 2 3 4 5 6))` (parentheses added to show AST) or any of the values in between. In the particular case of primitives which are already built-in with a known return type (such as `list`) we can check to make sure that they are not of the form `A* => (B => C)` (eg. they do not have return type ReporterTask), but this isn't necessarily generalizable and one could easily make an extension primitive which would have strange behavior in the corner case.

**Comment by qiemem** [comment]
Thanks @mrerrormessage. The thing is, NetLogo doesn't even attempt to determine how to distribute arguments. Without parenthesis, it always passes them to the HOF. That is, `(map filter-list 1 2 3 4 5 6)` is always interpreted as `(map (filter-list) 1 2 3 4 5 6)` (or rather `(map task [ filter-list ?1] 1 2 3 4 5 6)` assuming the default arity of `filter-list` is 1). The only way to get it to do `(map (filter-list 1) 2 3 4 5 6)` (or whatever) is with full parenthesization.

That's what I was trying to say with:

> More specifically, a sequence of tokens following a task-taking primitive will be parsed as a single expression if and only if they are surrounded by parentheses

Thus, theoretically, you shouldn't have to fix the arity of the task.
